### PR TITLE
Add QANet model.

### PR DESCRIPTION
**Pull request by qyhfbqz**
## Description ##
Generally, we have achieved the state-of-the-art performance (F1: 79.5 EM: 70.2) on SQuAD 1.1 dataset.

1. Add data preprocessing including specific bucketing, batchify, etc to be aligned with TF implementation.
2. Add QANet including character and word embedding layer, embedding block, encoder block(e.g. position encoder, depthwise conv, self-attention, positionwise ffn), output layer.
3. Add attention layer including the trilinear similarity computation, etc.
4. Add official evaluation script to report the F1 and EM score.
5. Add other utilities in the training including Exponential Moving Average, Warmup, etc.

### CODE REVIEW

**Comment by cgraywang** [comment]
Thanks for the great work done by Yan Feng @qyhfbqz !

**Comment by mli** [comment]
Job PR-339/14 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-339/14/index.html

File: scripts/question_answering/model.py [hunk 1]
```diff
@@ -1,636 +1,727 @@
...
+
+import mxnet as mx
+from mxnet import gluon, nd
+from mxnet.initializer import MSRAPrelu, Normal, Uniform, Xavier
+from gluonnlp.model import (DotProductAttentionCell, Highway,
+                            MultiHeadAttentionCell)
+
+
+try:
+    from config import (LAYERS_DROPOUT, EMB_ENCODER_CONV_CHANNELS, NUM_HIGHWAY_LAYERS,
```
Comment (cgraywang) [inline scripts/question_answering/model.py]: I suggest we can use `from config import *` and use e.g. `config.LAYERS_DROPOUT` in the context. Similarly, we can use the same method for the other hyperparameters from the config in the other places.

File: scripts/question_answering/util.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+def zero_grad(params):
+    r"""
+    Set the grad to zero.
+    """
+    for _, paramter in params.items():
+        paramter.zero_grad()
+
+
+class ExponentialMovingAverage():
```
Comment (eric-haibin-lin) [inline scripts/question_answering/util.py]: This is generally very useful. We should consider moving this to MXNet/GluonNLP official API later @szha @cgraywang

File: scripts/question_answering/train.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+import mxnet as mx
+from mxnet import (autograd, gluon, nd)
+from tqdm import tqdm
+
+from data_loader import DataLoader
+from evaluate import evaluate
+
+try:
+    from config import (EPOCHS, SAVE_MODEL_PREFIX_NAME, SAVE_TRAINER_PREFIX_NAME,
```
[the remark this answers is not in the document]
Comment (cgraywang) [reply scripts/question_answering/train.py]: When running the script on mac locally and ec2, it seems that the import path is different.

**Comment by mli** [comment]
Job PR-339/15 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-339/15/index.html

File: scripts/question_answering/train.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+        batch_sizes = context.shape[0]
+        context = gluon.utils.split_and_load(
+            data=context,
+            ctx_list=CTX,
+            # even_split=False
+        )
+        c_mask = gluon.utils.split_and_load(
+            data=c_mask,
+            ctx_list=CTX,
+            # even_split=False
```
[the remark this answers is not in the document]
Comment (qyhfbqz) [reply scripts/question_answering/train.py]: Done.

**Comment by mli** [comment]
Job PR-339/19 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-339/19/index.html

**Comment by szha** [comment]
Thanks for the efforts in making it happen! @qyhfbqz @cgraywang

**Pull request merged**